### PR TITLE
Add several new watchers

### PR DESCRIPTION
**Pull request by zollman**
This PR extends the dynamic watchers & auditors framework from https://github.com/Netflix/security_monkey/pull/405 to support several new watchers.

List of added watchers:
    * CloudTrail
    * AWSConfig
    * AWSConfigRecorder
    * DirectConnect::Connection
    * DirectConnect::VirtualGateway
    * EC2::EbsSnapshot
    * EC2::EbsVolume
    * EC2::Image
    * EC2::Instance
    * ENI
    * KMS::Grant
    * KMS::Key
    * Lambda
    * RDS::ClusterSnapshot
    * RDS::DBCluster
    * RDS::DBInstace
    * RDS::Snapshot
    * RDS::SubnetGroup
    * Route53
    * Route53Domains
    * VPC::DHCP
    * VPC::Endpoint
    * VPC::FlowLog
    * VPC::NatGateway
    * VPC::NetworkACL
    * VPC::Peering

Additional changes:
  * Move rds[security_group] to rds/ directory.
  * Update vpc/route_table to use boto3 lib.
  * Add tests for tech types supported by moto
  * Initialize name to avoid UnboundLocalError
  * Update RDS watcher ephemeral values
  * Use boto3.session.get_available_regions in select watchers
  * Convert routetable watcher to decorator pattern
  * Convert route53 domains to decorator pattern
  * Handle the case where the aws principle is a string
  * Catching assume role exception in decorator

### CODE REVIEW

**Comment by monkeysecurity** [comment]
Do we have the list of new IAM permissions that will be required for all these new watchers?

On first run I get this:

> 2016-09-12 23:18:29,487 INFO: 'Problem Connecting to peering/demo_account/None:\nAn error occurred (UnauthorizedOperation) when calling the DescribeVpcPeeringConnections operation: You are not authorized to perform this operation.'

**Comment by zollman** [comment]
The specific error you're seeing should've been addressed by the setup script already in develop - https://github.com/Netflix/security_monkey/blob/develop/scripts/secmonkey_role_setup.py#L62 .

We've changed our deployment process to separate the role setup into a different provisioning script, so I'll paste it here but we could bake it into `secmonkey_role_setup.py` if you'd rather us push a new commit to the PR. We use the managed SecurityAudit policy plus a delta, attached as an inline policy to the role SecurityMonkey uses:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "acm:ListCertificates",
                "acm:DescribeCertificate",
                "es:ListDomainNames",
                "es:DescribeElasticsearchDomainConfig",
                "lambda:ListFunctions",
                "ses:ListVerifiedEmailAddresses",
                "ses:GetIdentityVerificationAttributes",
                "sns:ListSubscriptionsByTopic"
            ],
            "Resource": [
                "*"
            ]
        }
    ]
}
```

File: security_monkey/watchers/kms.py [hunk 1]
```diff
@@ -111,84 +129,76 @@ def slurp(self):
...
+
+                            if config[u"Aliases"]:
+                                alias = config[u"Aliases"][0]
+                                alias = alias[len('alias/'):]  # Turn alias/name into just name
+                            else:
+                                alias = "[No Aliases]"
+
+                            name = "{alias} ({key_id})".format(alias=alias, key_id=key_id)
+
+                            item = KMSMasterKey(region=kwargs['account_name'], account=kwargs['account_name'], name=name, arn=config.get('Arn'), config=dict(config))
```
Comment (monkeysecurity) [inline security_monkey/watchers/kms.py]: bug: `region=kwargs['account_name']`

File: security_monkey/watchers/vpc/subnet.py [hunk 2]
```diff
@@ -102,10 +95,14 @@ def slurp(self):
                         "vpc_id": subnet.vpc_id
                     }
 
-                    item = SubnetItem(region=region.name, account=account, name=subnet_name, arn=arn, config=config)
+                    item = SubnetItem(region=kwargs['region'],
+                                      account=kwargs['account_name'],
+                                      name=subnet_name, config=config)
```
Comment (monkeysecurity) [inline security_monkey/watchers/vpc/subnet.py]: Dropped `arn=arn`

File: security_monkey/watchers/vpc/route_table.py [hunk 3]
```diff
@@ -38,86 +33,95 @@ class RouteTable(Watcher):
...
-                        "vpc_id": route_table.vpc_id,
+                        "tags": joined_tags,
+                        "vpc_id": route_table.get('VpcId'),
                         "associations": associations
                     }
 
-                    item = RouteTableItem(region=region.name, account=account, name=subnet_name, arn=arn, config=config)
+                    item = RouteTableItem(region=kwargs['region'],
+                                          account=kwargs['account_name'],
+                                          name=subnet_name, config=config)
```
Comment (monkeysecurity) [inline security_monkey/watchers/vpc/route_table.py]: dropped `arn=arn`

File: security_monkey/watchers/config_recorder.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+from security_monkey import app
+from boto.ec2 import regions
+
+
+class ConfigRecorder(Watcher):
+    index = 'configrecorder'
+    i_am_singular = 'Config Recorder'
+    i_am_plural = 'Config Recorders'
+    # TODO: Replace hardcoded region with `get_available_regions` after next boto3 release
+    # Specific PR here: https://github.com/boto/boto3/pull/531/files
```
Comment (monkeysecurity) [inline security_monkey/watchers/config_recorder.py]: boto3/531 has been merged. Do we know if it's been included in a release? (Can this TODO go away yet?)

File: security_monkey/watchers/config.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+"""
+from security_monkey.watcher import Watcher
+from security_monkey.watcher import ChangeItem
+from security_monkey.constants import TROUBLE_REGIONS
+from security_monkey.exceptions import BotoConnectionIssue
+from security_monkey import app
+from boto.rds import regions
+
+AVAILABLE_REGIONS = ['us-east-1']
```
Comment (monkeysecurity) [inline security_monkey/watchers/config.py]: Is this an AWS limitation?

File: security_monkey/watchers/direct_connect/virtual_gateway.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        item_list = []
+        exception_map = {}
+        for account in self.accounts:
+            for region in regions():
+                app.logger.debug(
+                    "Checking {}/{}/{}".format(self.index, account, region.name))
+                try:
+                    dc = connect(account, 'boto3.ec2.client', region=region)
+                    response = self.wrap_aws_rate_limited_call(
+                        dc.describe_vpn_gateways
```
Comment (monkeysecurity) [inline security_monkey/watchers/direct_connect/virtual_gateway.py]: The file direct_connect/virtual_gateway.py:
- does not call `describe_virtual_gateways()` as you might expect with a 'directconnect' boto3 client
- instead calls `describe_vpn_gateways()` with an `ec2` boto3 client

Should probably be renamed to vpn_gateways.py and maybe placed in the ec2 or vpc folder.

**Comment by aebie** [comment]
@monkeysecurity should submit the fixes above and resubmit? There is also an additional bug we have found internally and fixed since this PR was made

**Comment by monkeysecurity** [comment]
Hey @aebie - if you just want to push a new commit to your `new-watchers` branch, I can rebase that on top of the other changes I've made.

**Comment by monkeysecurity** [comment]
Hey @aebie,

Which items were you thinking about updating?

- direct_connect/virtual_gateway.py
- boto3/531 TODO
- config `AVAILABLE_REGIONS`

I've already fixed the `region=kwargs['account_name']` in `kms.py` and added ARN to all applicable watchers in my `Bridgewater-new-watchers` branch: https://github.com/Netflix/security_monkey/commits/Bridgewater-new-watchers

**Comment by aebie** [comment]
@monkeysecurity I have a fix for the TODO/531 - changing to the updated iter_account_region decorator tag which involves encapsulation of get_account_regions in the decorator
The other fix is problems we found on our side with the new decorator that cause logging to be messed up and find_changes not to run multiple accounts. 

Hopefully Aaron will push these changes today.

**Comment by zollman** [comment]
pushed the decorator tag fixes for cherry-pick to #411

**Comment by monkeysecurity** [comment]
Thanks @zollman.  I merged the update to `decorators.py` and `config_recorder.py` into my Bridgewater-new-watchers branch.

SO I guess my last question before merging is whether `direct_connect/virtual_gateway.py` is misnamed.

> does not call describe_virtual_gateways() as you might expect with a 'directconnect' boto3 client
instead calls describe_vpn_gateways() with an ec2 boto3 client

**Comment by aebie** [comment]
@monkeysecurity are you suggesting we simply rename an move the virtual_gateway file?

**Comment by monkeysecurity** [comment]
Hey @aebie,

I'm not super familiar with those services so I just wanted to bring up what looked like a problem and get your opinion on what should be done.  Are you saying you think it's okay the way it is?

**Comment by aebie** [comment]
@monkeysecurity To be honest I didn't write this watcher and the guy who did has moved on to another project so I am not sure what this thinking behind the placement was. I would guess that the thought was that VPN was similar to direct connect in in concept so the two should be grouped together. I can also see the logic of grouping it with the vpc especially since that's where it lives in the AWS console. Personally I'd like to align with the AWS console so would be happy to move it and rename the watcher Virtual Private Gateway or something to the effect.

**Comment by monkeysecurity** [comment]
Awesome. Thanks @aebie .

**Comment by monkeysecurity** [comment]
I'm going to remove `direct_connect/virtual_gateway.py` from this PR so I can merge it.  We can add it back later with a different PR.

**Comment by monkeysecurity** [comment]
Merged in #411

**Pull request closed without merging**